Added s2n_client_hello_get_session_id calls (#3006)
aws/s2n-tls: docs/USAGE-GUIDE.md

**s2n_client_hello_get_extension_length** returns the number of bytes the given extension type takes on the ClientHello message received by the server; it can be used to allocate the **out** buffer.
**s2n_client_hello_get_extension_by_id** copies into the **out** buffer **max_length** bytes of a given extension type on the ClientHello and returns the number of copied bytes.

### s2n\_client\_hello\_get\_session\_id

```c
int s2n_client_hello_get_session_id_length(struct s2n_client_hello *ch, uint32_t *out_length);
int s2n_client_hello_get_session_id(struct s2n_client_hello *ch, uint8_t *out, uint32_t *out_length, uint32_t max_length);
```

> These functions allow to get the session id as sent by the client in ClientHello message. The session id on the **s2n_connection** may change later when the server sends ServerHello, see **s2n_connection_get_session_id** on how to get the final session id on the connection.

Reviewer: ```suggestion
These functions retrieve the session id as sent by the client in the ClientHello message. The session id on the **s2n_connection** may change later when the server sends the ServerHello; see **s2n_connection_get_session_id** for how to get the final session id used for future session resumption.
```

**s2n_client_hello_get_extension_length** returns the number of bytes the given extension type takes on the ClientHello message received by the server; it can be used to allocate the **out** buffer.
**s2n_client_hello_get_extension_by_id** copies into the **out** buffer **max_length** bytes of a given extension type on the ClientHello and returns the number of copied bytes.

### s2n\_client\_hello\_get\_session\_id

```c
ssize_t s2n_client_hello_get_session_id_length(struct s2n_client_hello *ch);
ssize_t s2n_client_hello_get_session_id(struct s2n_client_hello *ch, uint8_t *out, uint32_t max_length);
```

- **ch** The s2n_client_hello on the s2n_connection. The handle can be obtained using **s2n_connection_get_client_hello**.
- **out** Pointer to a buffer into which the session id bytes should be copied.
- **max_length** Max number of bytes to copy into the **out** buffer.

> **s2n_client_hello_get_session_id_length** returns the number of bytes the session id takes on the ClientHello message received by the server; it can be used to allocate the **out** buffer.

Reviewer: ```suggestion
**s2n_client_hello_get_session_id_length** returns the length of the ClientHello session id in bytes. It can be used to allocate the **out** buffer.
```

> **s2n_client_hello_get_session_id** copies into the **out** buffer **max_length** bytes of a session id on the ClientHello and returns the number of copied bytes.

Reviewer: ```suggestion
**s2n_client_hello_get_session_id** copies **max_length** bytes of a ClientHello session id into the **out** buffer and returns the number of copied bytes.
```

> 
> Note: the result is can be different from **s2n_connection_get_session_id**, as session id may change with ServerHello which will send a different session id if we're not resuming the ClientHello provided session.

Reviewer: It might be worth making this more prominent than a note. "Which of these two APIs do I use" will probably be the main question someone reading this documentation has.

```suggestion

Note: the result can be different from **s2n_connection_get_session_id**, as session id may change if the server chooses not to resume the provided session and instead responds to the ClientHello with a new session id.
```

> 

Author: Right, mainly because I'd like to call it after the handshake was done and not bother adding the callback.

**s2n_client_hello_get_extension_length** returns the number of bytes the given extension type takes on the ClientHello message received by the server; it can be used to allocate the **out** buffer.
**s2n_client_hello_get_extension_by_id** copies into the **out** buffer **max_length** bytes of a given extension type on the ClientHello and returns the number of copied bytes.

### s2n\_client\_hello\_get\_session\_id

```c
int s2n_client_hello_get_session_id_length(struct s2n_client_hello *ch, uint32_t *out_length);
int s2n_client_hello_get_session_id(struct s2n_client_hello *ch, uint8_t *out, uint32_t *out_length, uint32_t max_length);
```

- **ch** The s2n_client_hello on the s2n_connection. The handle can be obtained using **s2n_connection_get_client_hello**.
- **out** Pointer to a buffer into which the session id bytes should be copied.
- **out_length** Pointer to a **uint32_t** which for **s2n_client_hello_get_session_id_length** will be used to store the full length of session_id and for **s2n_client_hello_get_session_id** will be used to store the number of bytes copied to the **out** buffer.

> - **max_length** Max number of bytes to copy into the **out** buffer.

Reviewer: Not all of the other documentation entries discuss arguments separately from the methods. Since out_length has two different uses, is it worth trying to discuss both at once? This section might not be helpful

Author: Fair enough - followed previous docs too much. I'll split it up.

Author: Got rid of the list and moved into description of individual function.

> **s2n_client_hello_get_session_id_length** in case of success returns **S2N_SUCCESS** and stores in **out_length** the length of the ClientHello session id in bytes; in case of failure returns **S2N_FAILURE** with no guarantees around **out_length** content. The **out_length** can be used to allocate the **out** buffer for the **s2n_client_hello_get_session_id** call.
> **s2n_client_hello_get_session_id** in case of success returns **S2N_SUCCESS**, copies up to **max_length** bytes of a ClientHello session_id into the **out** buffer and stores the number of copied bytes in **out_length**; in case of failure returns **S2N_FAILURE** with no guarantees around **out** and **out_length** content.

Reviewer: ```suggestion
**s2n_client_hello_get_session_id_length** stores the ClientHello session id length in bytes in **out_length**. The **out_length** can be used to allocate the **out** buffer for the **s2n_client_hello_get_session_id** call.
**s2n_client_hello_get_session_id** copies up to **max_length** bytes of the ClientHello session_id into the **out** buffer and stores the number of copied bytes in **out_length**.
```

We generally don't discuss standard error handling for every method. It's covered in https://github.com/aws/s2n-tls/blob/12720117108630907377eedfe79de7e138c0e107/docs/USAGE-GUIDE.md#error-handling

Author: Writing docs is hard (: Thanks, updated.

Note: the result can be different from **s2n_connection_get_session_id**, as session id may change if the server chooses not to resume the provided session and instead responds to the ClientHello with a new session id.

> 

Reviewer: I [still](https://github.com/aws/s2n-tls/pull/3006/files/f4f5cd0c3d9cd754510ffee2315709537fc214fe#r687973454) think this should be more prominent than a note. The difference in behaviors is the whole point of the API.

Author: I'll try to move it out of the note and update the `s2n_connection_get_session_id` doc too.

Author: Reworded it a bit, let me know what you think.

**s2n_connection_get_session_length** returns number of bytes needed to store serialized session state; it can be used to allocate the **session** buffer.

> **s2n_connection_get_session_id_length** returns latest session id length from the connection. Session id length will be 0 for TLS versions >= TLS1.3 as stateful session resumption has not yet been implemented in TLS1.3.
> 

Reviewer: ```suggestion
**s2n_connection_get_session_id_length** returns the latest session id length from the connection. Session id length will be 0 for TLS versions >= TLS1.3 as stateful session resumption has not yet been implemented in TLS1.3.

```

> **s2n_connection_get_session_id** get the latest session id from the connection and copies into the **session_id** buffer and returns the number of copied bytes. The session id may change between s2n receiving ClientHello and sending ServerHello, the function will always provide information on the latest session id. See **s2n_client_hello_get_session_id** function to get session id as it was sent by the client in ClientHello message.

Reviewer: ```suggestion
**s2n_connection_get_session_id** gets the latest session id from the connection, copies it into the **session_id** buffer, and returns the number of copied bytes. The session id may change between s2n receiving the ClientHello and sending the ServerHello, but this function will always describe the latest session id. See **s2n_client_hello_get_session_id** to get the session id as it was sent by the client in the ClientHello message.
```
